src/Model/LazyBSONDocument.php

```diff
@@ -0,0 +1,254 @@
+<?php
+/*
+ * Copyright 2016-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use AppendIterator;
+use ArrayAccess;
+use ArrayIterator;
+use CallbackFilterIterator;
+use Iterator;
+use IteratorAggregate;
+use JsonSerializable;
+use MongoDB\BSON\Document;
+use MongoDB\BSON\Serializable;
+use MongoDB\Codec\CodecLibrary;
+use MongoDB\Codec\LazyBSONArrayCodec;
+use MongoDB\Codec\LazyBSONDocumentCodec;
+use MongoDB\Exception\InvalidArgumentException;
+use ReturnTypeWillChange;
+
+use function array_key_exists;
+use function is_array;
+use function is_object;
+use function iterator_to_array;
+use function sprintf;
+use function trigger_error;
+
+use const E_USER_WARNING;
+
+/**
+ * Model class for a BSON document.
+ *
+ * The internal data will be cast to an object during BSON serialization to
+ * ensure that it becomes a BSON document.
+ *
+ * @api
+ */
+class LazyBSONDocument implements ArrayAccess, IteratorAggregate, JsonSerializable, Serializable
+{
+    /** @var CodecLibrary|null */
+    private $library;
+
+    /** @var Document */
+    private $bson;
+
+    /** @var array<string, mixed> */
+    private $read = [];
+
+    /** @var array<string, true> */
+    private $notFound = [];
+
+    /** @var array<string, mixed> */
+    private $set = [];
+
+    /** @var array<string, true> */
+    private $unset = [];
+
+    /**
+     * Deep clone this lazy document.
+     */
+    public function __clone()
+    {
+        $this->bson = clone $this->bson;
+
+        foreach ($this->set as $key => $value) {
+            if (is_object($value)) {
+                $this->set[$key] = clone $value;
+            }
```

Comment:
Similar to my earlier comment in LazyBSONArray, I think you may want `recursive_copy()` here. And perhaps a regression test for both that fails with this original code.